File: src/harness/python/tools/bash.py
"""The bash tool: run one shell command.

This is part of the tools seam (a behavior seam).
The tool calls env, never the machine directly.
"""

from env.base import Env, EnvError
from tools.registry import env_error, text

TIMEOUT_SECONDS = 30


def bash(env: Env, messages: dict, arguments: dict) -> str:
    """Run the command. Return stdout, then stderr, then the exit code line."""
    try:
        stdout, stderr, code = env.run(arguments["command"], TIMEOUT_SECONDS)
    except EnvError as error:
        return env_error(messages, error)
    return with_newline(stdout) + with_newline(stderr) + text(messages, "exit_code", n=code)


def with_newline(part: str) -> str:
    """Return the part with one newline at the end. An empty part stays empty."""
    if part == "" or part.endswith("\n"):
        return part
    return part + "\n"

File: src/harness/typescript/tools/bash.ts
// The bash tool: run one shell command.
//
// This is part of the tools seam (a behavior seam).
// The tool calls env, never the machine directly.

import { EnvError, type Env } from "../env/base.ts";
import { envError, text, type Arguments, type Messages } from "./registry.ts";

const TIMEOUT_SECONDS = 30;

// Run the command. Return stdout, then stderr, then the exit code line.
export async function bash(env: Env, messages: Messages, args: Arguments): Promise<string> {
  let stdout: string;
  let stderr: string;
  let code: number;
  try {
    ({ stdout, stderr, code } = await env.run(args.command as string, TIMEOUT_SECONDS));
  } catch (error) {
    if (error instanceof EnvError) return envError(messages, error);
    throw error;
  }
  return withNewline(stdout) + withNewline(stderr) + text(messages, "exit_code", { n: code });
}

// Return the part with one newline at the end. An empty part stays empty.
function withNewline(part: string): string {
  if (part === "" || part.endsWith("\n")) return part;
  return part + "\n";
}
